lib/Dialect/DC/Transforms/DCPrintDot.cpp

```diff
@@ -0,0 +1,338 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Analysis pass.
+//
+//===----------------------------------------------------------------------===//
+#include "circt/Dialect/DC/DCOps.h"
+#include "circt/Dialect/DC/DCPasses.h"
+#include "mlir/Pass/Pass.h"
+
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/DC/DCOps.h"
+#include "mlir/Support/IndentedOstream.h"
+#include "llvm/ADT/ScopedHashTable.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace circt {
+namespace dc {
+#define GEN_PASS_DEF_DCDOTPRINT
+#include "circt/Dialect/DC/DCPasses.h.inc"
+} // namespace dc
+} // namespace circt
+
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+using ValueMap = llvm::ScopedHashTable<mlir::Value, std::string>;
+
+namespace {
+
+/// all Comb and DC nodetypes
+enum NodeType {
+  addOp,
+  andOp,
+  xorOp,
+  orOp,
+  mulOp,
+  muxOp,
+  eqOp,
+  neOp,
+  gtOp,
+  geOp,
+  ltOp,
+  leOp,
+  branchOp,
+  bufferOp,
+  forkOp,
+  fromESIOp,
+  joinOp,
+  mergeOp,
+  packOp,
+  selectOp,
+  sinkOp,
+  sourceOp,
+  toESIOp,
+  unpackOp,
+  null
+};
+
+std::string stringify(NodeType type) {
+  switch (type) {
+  case addOp:
+    return "+";
+  case andOp:
+    return "&&";
+  case xorOp:
+    return "^";
+  case orOp:
+    return "||";
+  case mulOp:
+    return "x";
+  case muxOp:
+    return "mux";
+  case eqOp:
+    return "==";
+  case neOp:
+    return "!=";
+  case gtOp:
+    return ">";
+  case geOp:
+    return ">=";
+  case ltOp:
+    return "<";
+  case leOp:
+    return "<=";
+  case branchOp:
+    return "branch";
+  case bufferOp:
+    return "buffer";
+  case forkOp:
+    return "fork";
+  case fromESIOp:
+    return "fromESI";
+  case joinOp:
+    return "join";
+  case mergeOp:
+    return "merge";
+  case packOp:
+    return "pack";
+  case selectOp:
+    return "select";
+  case sinkOp:
+    return "sink";
+  case sourceOp:
+    return "source";
+  case toESIOp:
+    return "toESI";
+  case unpackOp:
+    return "unpack";
+  case null:
+    return "null";
+  default:
+    return "Unknown NodeType";
+  }
+}
+
+/// stores operand and results for each node in the dot graph
+struct DotNode {
+  NodeType nodeType;
+  llvm::SmallVector<std::pair<mlir::Value, std::string>> incoming;
+  llvm::SmallVector<std::pair<mlir::Value, std::string>> outgoing;
+};
+} // namespace
+
+/// gives a unique name to each value in the graph
+llvm::SmallVector<std::pair<mlir::Value, std::string>> static valueToName(
+    const llvm::SmallVector<mlir::Value> &values,
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> &currentMap,
+    bool tokenFlag) {
+  SmallVector<std::pair<mlir::Value, std::string>> res;
+  for (auto [i, v] : llvm::enumerate(values)) {
+    auto found = false;
+    for (const auto &cm : currentMap) {
```

Comment:
Nit: you can probably use structured bindings as well here.
```suggestion
    for (const [key, value] &cm : currentMap) {
```